Add JSON button to trace page

virtuald opened the pull request:
Well, since the "[privileged minority](https://github.com/openzipkin/zipkin/issues/1093#issuecomment-213216090)" hasn't done it yet, I figured I would since I keep finding that I need it. :) 
- Shows prettified JSON in modal dialog
- Fixes #1060

I'm very new to the various frameworks that the UI is using, so I'm open to suggestions if there's anything that I did that was a bit wonky.

eirslett commented on zipkin-ui/js/component_ui/jsonPanel.js:
```diff
@@ -0,0 +1,16 @@
+import {component} from 'flightjs';
+import $ from 'jquery';
```
Defined, but never used (try running `npm run lint`)

eirslett commented on zipkin-ui/js/component_ui/traceToMustache.js:
```diff
@@ -87,7 +87,9 @@ export function formatEndpoint({ipv4, port = 0, serviceName = ''}) {
 }
 
 export default function traceToMustache(trace) {
+  const raw = trace;
```
The raw data doesn't belong inside the modelview; this function is supposed to transform the incoming raw data into a format that is ready for mustache rendering.
Can you try refactoring, so that raw is not passed as part of the modelview?

eirslett commented on zipkin-ui/js/page/trace.js:
```diff
@@ -24,9 +25,10 @@ const TracePageComponent = component(function TracePage() {
       ServiceFilterSearchUI.attachTo('#serviceFilterSearch');
       SpanPanelUI.attachTo('#spanPanel');
-      TraceUI.attachTo('#trace-container');
+      TraceUI.attachTo('#trace-container', {raw: modelview.raw});
```
The parameter that is called "modelview" today could be renamed to "data", and then "data" would contain two properties: "modelview" and "raw"?

eirslett commented:
I had some comments; otherwise, great work!

codefromthecrypt commented:
Thanks for the effort and review folks. This is very popular!

virtuald commented:
Thanks for the comments, I'll address these tonight or tomorrow.

virtuald committed:
Add JSON button to trace page

- Shows prettified JSON in modal dialog
- Adds download link for JSON
- Fixes #1060

virtuald commented:
Comments addressed.

codefromthecrypt commented:
running it locally and will merge (and release) after!

virtuald commented:
I've got another set of changes coming too dealing with better error messages, just testing it now.

codefromthecrypt commented:
ah ok thx for the heads-up. this is a different PR, right?

virtuald commented:
Yep. Different PR.

codefromthecrypt commented on zipkin-ui/js/component_ui/trace.js:
```diff
@@ -443,7 +448,11 @@ export default component(function trace() {
+    this.on(document, 'uiShowJson', this.showJson);
+
     const self = this;
+    self.raw = data.raw;
```
how is this controlled?

virtuald replied:
It comes through the trigger in `js/component_data/trace.js`

codefromthecrypt replied:
sorry I meant how do you control this via the UI. Ex if the user wants to
view the raw json (api/v1/trace/id?raw), what do they do? Or does raw mean
something else?

virtuald replied:
This PR does not provide a mechanism to do that. I suspect that is best saved for future work -- if there isn't a way to show an uploaded trace, no sense downloading the trace? 

I could be convinced otherwise, it would be easy to add an actual raw download link somewhere.

codefromthecrypt replied:
"raw" in the code here confused me as it is a parameter GET /api/v1/trace
api

virtuald replied:
Oh, ha. Didn't realize that.

codefromthecrypt replied:
ex not sure why the website doesn't render it, but here's the definition
- name: raw
  in: query
  required: false
  description: |
  Note this flag has no value. Ex. /trace/{traceId}?raw
  Normally, the trace endpoint cleans trace data. For example, it merges
  spans by id, adds missing timestamp or duration, corrects clock skew..
  Specifying this flag is a debug case, when you are debugging zipkin
  logic or zipkin instrumentation, and want to see the input to these
  adjusters. For example, this might explain or rule out clock skew.
  type: boolean

https://github.com/openzipkin/zipkin-api/blob/master/zipkin-api.yaml#L146

virtuald replied:
Interesting. I suppose we could add a mechanism that could display either. I chose the default since it was already retrieved by the server -- so it would show the JSON that's currently being displayed without another call to the server. Theoretically, if you made another call down, the JSON could show something different which wasn't currently being displayed.

codefromthecrypt replied:
good point.

Also, principle of least surprise wins for not defaulting to "raw" flag:
without disclaimers, it would confuse people when timestamps don't match
what the UI says.

I'm game for as-is.. I'd prefer that variable to be named something besides
raw (or clarified that it doesn't imply the raw parameter)

virtuald replied:
Renamed to `traceJson`

codefromthecrypt commented:
fyi here are some screen shots:

![screen shot 2016-05-31 at 10 49 49 am](https://cloud.githubusercontent.com/assets/64215/15661777/74a7c538-271d-11e6-92cc-05466f9443c2.png)

![screen shot 2016-05-31 at 10 49 36 am](https://cloud.githubusercontent.com/assets/64215/15661778/77016276-271d-11e6-855a-7433a15e93df.png)

yurishkuro commented:
is json raw or merged?

virtuald commented:
The JSON should be whatever was retrieved from the call to `/api/v1/traces/XXX`, as it's the same object.

codefromthecrypt commented:
@yurishkuro you have this feature already.. do you default to raw json? any thoughts about what the default should be?

yurishkuro commented:
@adriancole  we added two buttons, raw and regular json, as both come in handy. But we just have them as plain links to json documents, not popups, and we leave pretty rendering to browser plugin. That way it's easy to search inside the json, and do a Save As.

codefromthecrypt commented:
interesting. search works fine with the pop-up, but save-as doesn't. It is
a common support case... @virtuald https://github.com/virtuald thoughts
on save-as etc?

codefromthecrypt commented:
![screen shot 2016-05-31 at 10 49 36 am](https://cloud.githubusercontent.com/assets/64215/15663863/f5c63154-2731-11e6-9fd4-364961712816.png)

Here's a thought. you can add a save or download link to the pop-up. this could point to /api/v1/trace/id for example.

eirslett commented:
+1 for adriancole's suggestion!

virtuald commented:
I added a save icon, and also the ability to right click on the JSON button and do a save as (as that's an intuitive option anyways).

eirslett commented on zipkin-ui/js/component_ui/traceToMustache.js:
```diff
@@ -167,6 +168,7 @@ export default function traceToMustache(trace) {
   const spansBackup = spans;
 
   return {
+    traceId,
```
I guess this isn't needed anymore?

virtuald replied:
It's used in trace.mustache line 7.

eirslett replied:
Ah, I see it now! Sorry.

eirslett commented on zipkin-ui/templates/trace.mustache:
```diff
@@ -4,13 +4,14 @@
     <li class=''><a href='#'><strong>Services:</strong> <span class='badge'>{{services}}</span></a></li>
     <li class=''><a href='#'><strong>Depth:</strong> <span class='badge'>{{depth}}</span></a></li>
     <li class=''><a href='#'><strong>Total Spans:</strong> <span class='badge'>{{totalSpans}}</span></a></li>
+    <li class='navbar-right'><a href='/api/v1/trace/{{ traceId }}' id='traceJsonLink'><button type="button" class="btn btn-primary btn-xs badge" value="uiShowJson">JSON</button></a></li>
```
You don't need `<button>`here, you can just put the classes on the A directly, like this: `<a class="btn btn-primary btn-xs badge">` (I think)

virtuald replied:
Hm, can you use the `value` thing on an `<a>` element? Using the same event mechanism as everything else is convenient.

eirslett replied:
No, it doesn't have the value property. How about using an explicit click handler on the a element? I think that's cleaner overall.

virtuald replied:
Ok, done.

eirslett commented on zipkin-ui/templates/trace.mustache:
```diff
@@ -4,13 +4,14 @@
       <button type='button' class='btn btn-default' value='uiExpandAllSpans'>Expand All</button>
       <button type='button' class='btn btn-default active' value='uiCollapseAllSpans'>Collapse All</button>
-    </div>
+    </div> 
```
This extra space isn't needed

virtuald commented on zipkin-ui/js/page/trace.js:
```diff
@@ -17,19 +18,27 @@ const TracePageComponent = component(function TracePage() {
       FilterLabelUI.attachTo('.service-filter-label');
       ZoomOut.attachTo('#zoomOutSpans');
 
+      this.$node.find('#traceJsonLink').click(e => {
```
@eirslett you're right, this ends up being cleaner anyways, it never really belonged in trace.js

eirslett commented:
Otherwise, LGTM! ship it

codefromthecrypt commented:
yay json button!
